vignettes/motivation.Rmd

```diff
@@ -0,0 +1,67 @@
+---
+title: "Motivation for S7"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Motivation for S7}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+R already has two OO systems built-in (S3 and S4) and many additional OO systems are available in CRAN packages.
+Why did we decide another system (S7) was necessary?
+This vignette will discuss some of the motivations, focussing on the aspects of S3 and S4 that have been found to be particularly challenging in practice.
+
+```{r setup}
+library(S7)
+```
+
+## Challenges with S3
+
+-   S3 is very informal, meaning that there's no formal definition of a class.
+    This makes it impossible to know exactly which properties an object should or could possess, or even what it's parent class should be.
+    S7 resolves this problem with a formal definition encoded in a class object produced by `new_class()`.
+    This includes support for validation (and avoiding validation where needed) as inspired by S4.
+
+-   When a new user encounter an S3 generic, they are often confused because the implementation of the function appears to be missing.
+    S7 has a thoughtfully designed print method that makes it clear what methods are available and how to find their source code.
+
+-   Properties of an S3 class are usually stored in attributes, but, by default, `attr()` does partial matching, which can lead to bugs that are hard to diagnose.
+    Additionally, `attr()` returns `NULL` if an attribute exists, so misspelling an attribute can lead to subtle bugs.
+    `@` fixes both of these problems.
+
+-   S3 method dispatch is complicated for compatibility with S.
+    This complexity affects relatively little code, but when you attempt to dive into the details it makes `UseMethod()` hard to understand.
+    As much as possible, S7 avoids any "funny" business with environments or promises, so that there is no distinction between argument values and local values.
+
+-   `NextMethod()` is unpredictable you can't tell exactly what method will be called by only reading the code; you instead need to know both the complete class hierarchy and what other methods are currently registered (and loading a package might change those methods).
```

Comment:
```suggestion
-   `NextMethod()` is unpredictable since you can't tell exactly which method will be called by only reading the code; you instead need to know both the complete class hierarchy and what other methods are currently registered (and loading a package might change those methods).
```